[sooyitao] iP

sooyitao opened the pull request:
# HanaBot
> “Productivity is never an accident. It is always the result of a commitment to excellence, intelligent planning, and focused effort.” – [Paul J. Meyer](https://en.wikipedia.org/wiki/Paul_J._Meyer)

**HanaBot** is a chatbot that helps you organize your tasks efficiently. It's:
- Simple and intuitive to use
- Lightweight and ~~fast~~ _SUPER FAST_
- Command-line based for simplicity

### Getting Started
1. [Download HanaBot .jar file](https://github.com/sooyitao/ip/releases/tag/A-Jar)
2. `cd` into the folder of the downloaded jar file
3. Run `java -jar Hana.jar`
4. Start adding tasks to your task list! 🫡
5. Let it manage your tasks for you 😉

### What's New
- [x] Task Management: Add, edit, and remove tasks.
- [x] Deadlines: Set deadlines for your tasks.
- [ ] Recurring Tasks: Automate repetitive tasks (coming soon).
- [ ] Notifications: Receive reminders for upcoming tasks (coming soon).

**HanaBot** is also a great tool for Java developers to sharpen their skills. Here’s a snippet of the main entry point for the application:
``` java
public class Hana {
    public static void main(String[] args) {
        new Hana().run();
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

sooyitao committed:
Level-0

sooyitao committed:
Level-1

sooyitao committed:
Level 2. Add, List

sooyitao committed:
Level 3. Mark as Done

sooyitao committed:
Level 4. ToDos, Events, Deadlines

sooyitao committed:
Automated Text UI Testing

sooyitao committed:
Level 5. Handle Errors

sooyitao committed:
Level 6. Delete

sooyitao committed:
no message

sooyitao committed:
Add load save file

sooyitao committed:
Add save task

sooyitao committed:
no message

sooyitao committed:
Merge branch 'master' into branch-Level-7

# Conflicts:
#	src/main/java/Hana.java

sooyitao committed:
Revert "Merge branch 'master' into branch-Level-7"

This reverts commit 7f2d09def86fcc170d3d23a40bd1a363a24eb01d, reversing
changes made to 4c8b9e76e6d4b2bbf9ab41ec618a3769a8b6080f.

sooyitao committed:
Merge branch 'branch-Level-7'

sooyitao committed:
Add Dates and Times

sooyitao committed:
Add find by date

sooyitao committed:
Merge branch 'branch-Level-8'

sooyitao committed:
Updated to use more OOP

sooyitao committed:
Merge branch 'branch-A-MoreOOP'

sooyitao committed:
Divide classes into packages

sooyitao committed:
Merge branch 'branch-A-Packages'

sooyitao committed:
Merge branch 'add-gradle-support'

sooyitao committed:
Add JUnit Tests

sooyitao committed:
Merge branch 'branch-A-JUnit'

sooyitao committed:
Update build.gardle

sooyitao committed:
Add JavaDoc

sooyitao committed:
Merge branch 'branch-A-JavaDoc' into branch-A-CodingStandard

sooyitao committed:
Updated to comply with coding standard

sooyitao committed:
Add find by keyword command

sooyitao committed:
Merge branch 'branch-A-JavaDoc'

sooyitao committed:
Merge branch 'branch-A-CodingStandard'

sooyitao committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/hana/parser/Parser.java
#	src/main/java/hana/task/Task.java
#	src/main/java/hana/tasklist/TaskList.java
#	src/main/java/hana/ui/Ui.java

sooyitao committed:
Revert "Merge branch 'branch-Level-9'"

This reverts commit 43fa05a2f8fc0db73426d35103b3c77210f5b6b4, reversing
changes made to d0e7bbd826cfd4be64639f26943e3cac790a6f5a.

sooyitao committed:
Revert "Revert "Merge branch 'branch-Level-9'""

This reverts commit 1f9fc5fd143e4b3d4402933d5a9b3de525c0c843.

sooyitao committed:
Resolved conflicts

sooyitao committed:
Add JavaFX GUI for Hana Chatbot

* Implemented a graphical user interface (GUI) using JavaFX for the Hana chatbot application.
* Created a `MainWindow` class that sets up the main application window.
* Added `DialogBox` class for displaying messages with associated user and bot images.
* Integrated JavaFX FXML to layout the GUI components.
* Added functionality to handle user input and display bot responses in the chat interface.
* Updated `Hana` class to support GUI operations.
* Updated `ByeCommand` class to close application.
* Created `GreetingsCommand` to greet user at start of application.

sooyitao committed:
Merge branch 'Level-10'

sooyitao committed:
Update MainWindow

* Modifie MainWindow.java to use the fx:root construct
* Update MainWindow.fxml to remove explicit fx:controller declaration
* Add a Stage field to MainWindow class
* Introduce a setStage method in MainWindow to initialize and configure the Stage

sooyitao committed:
Merge branch 'Level-10'

Nimastic commented on README.md:
Looks good to me 👍

Nimastic reviewed (commented):
Looks good to me 😄

tohhengxing commented on src/main/java/hana/command/DeadlineCommand.java:
```diff
@@ -0,0 +1,58 @@
+        try {
+            String[] parts = input.substring(8).split(" /by ");
+            if (parts.length < 2) {
+                throw new HanaException("Deadline task must have a description and a due date.");
```
nit: can be a constant

kuiktzecheng commented on src/main/java/hana/command/EventCommand.java:
```diff
@@ -0,0 +1,60 @@
+            if (parts.length < 3) {
+                throw new HanaException("Event task must have a description, start time, and end time.");
+            }
+            DateTimeFormatter formatter = DateTimeFormatter.ofPattern("d/M/yyyy HHmm");
+            this.from = LocalDateTime.parse(parts[1].trim(), formatter);
+            this.to = LocalDateTime.parse(parts[2].trim(), formatter);
```
can consider to have 2 constants : INPUT_FORMATTER and OUTPUT_FORMATTER to format the input date and times

kuiktzecheng commented on src/main/java/hana/ui/Ui.java:
```diff
@@ -0,0 +1,138 @@
+    /**
+     * Prints all commands.
+     */
+    public void printCommands() {
```
Good method to have! To print out the string outputs based on what the user input is! Good idea!!

kuiktzecheng commented on src/main/java/hana/command/ByeCommand.java:
```diff
@@ -0,0 +1,44 @@
+    @Override
+    public void execute(TaskList taskList, Ui ui, Storage storage) throws HanaException {
+        ui.printBye();
+        PauseTransition pause = new PauseTransition(Duration.seconds(3));
+        pause.setOnFinished(event -> {
+            Platform.exit();
+        });
+        pause.play();
```
Is this necessary? May I know what the rationale is for?

sooyitao replied:
It creates a delay of 3 seconds before the app terminates so that the user can read the exit message rather than exit immediately after printing the message. It's not really necessary but without the delay, there's no point in printing the message since the user wouldn't have time to read it.

kuiktzecheng reviewed (commented):
Overall code looks good!!

sooyitao committed:
Add assertions to validate assumptions in Task, TaskList, and
Storage classes

Assertions were introduced to document and validate critical
assumptions about the state of the program at runtime. This ensures
that unexpected conditions are flagged early, aiding in debugging
and maintaining code correctness.

In TaskList, assertions check that the size of the task list
increases by 1 when a task is added, and decreases by 1 when a task
is deleted. This verifies the integrity of task list manipulation.

In Storage, assertions confirm that tasks are properly loaded and
saved, preventing potential issues with file I/O operations.

In Task, an assertion was added to ensure that task descriptions
are neither null nor empty upon creation, enforcing task data
validity.

These assertions enhance the robustness of the system by helping
catch logic errors and inconsistencies during development and
testing.

sooyitao committed:
Refactor codebase for improved error handling and readability

Parser Class: Refactored to use `switch` for better command handling
and flexibility.

Storage Class: Updated error messages and exception handling.

TaskList Class: Updated `Task` class with consistent naming
conventions and enhanced Javadoc comments.

DialogBox Class: Updated to comply with coding style

MainWindow Class: Removed unused code

These changes enhance maintainability, readability, and robustness
of the codebase.

sooyitao committed:
Merge pull request #1 from sooyitao/branch-A-Assertions

Add assertions to validate assumptions in Task, TaskList, and Storage classes

sooyitao committed:
Merge branch 'master' of https://github.com/sooyitao/ip

sooyitao committed:
Merge branch 'master' into branch-A-CodeQuality

sooyitao committed:
Set up continuous integration

sooyitao committed:
Revert "Set up continuous integration"

This reverts commit 331b811ffc92fc3fc5d8570466f80eece94cfbac.

sooyitao committed:
Merge pull request #2 from sooyitao/branch-A-CodeQuality

Refactor codebase for improved error handling and readability

sooyitao committed:
Merge branch 'master' of https://github.com/sooyitao/ip

sooyitao committed:
Add priorities to items

Users can mark task with priority level 1 (high) to 3 (low)
Users can list taks from high to low priority

sooyitao committed:
Bug Fix

Fixed minor javaFX bug in MainWindow

sooyitao committed:
Fix code quality

sooyitao committed:
Add Ui.png to doc

sooyitao committed:
Update README.md

sooyitao committed:
Update README.md
